hazelcast/src/main/java/com/hazelcast/internal/bootstrap/ClientNioAsyncReadHandler.java

```diff
@@ -79,10 +79,9 @@ private void loadConnection(ClientMessage message) {
     }
 
     private void consumeProtocolBytes(ByteBuffer buffer) {
-        StringBuffer sb = new StringBuffer();
-        for (int k = 0; k < 3; k++) {
-            sb.append((char) buffer.get());
-        }
+        // Note(sasha) : AFAIU, it's a trick to reduce garbage production. One object is better than 3.
+        StringBuilder sb = new StringBuilder();
+        sb.append((buffer.get() + buffer.get() + buffer.get()));
```

Comment:
Can't we do something like this to even not create the string builder?
```java
        upcast(buffer).position(buffer.position() + Protocols.PROTOCOL_LENGTH);

```

Reply:
This whole logic is flaky because we don't check the type of the client.